src/Microsoft.OData.Edm/ExtensionMethods/EdmTypeSemantics.cs

```diff
@@ -15,6 +15,15 @@ namespace Microsoft.OData.Edm
     /// </summary>
     public static class EdmTypeSemantics
     {
+
+        private static IDictionary<IEdmType, IDictionary<IEdmType, bool>> typeDictionaryCache;

```

Comment:
>IDictionary<IEdmType, IDictionary<IEdmType, bool>> [](start = 23, length = 50)

why is this not just an IDictionary<IEdmType, HashSet<IEdmType>> where the HashSet contains all of the types that the first type is or inherits from?

Reply:
For some it may not be inheritable , for some it is. we can only have hashset contain either. I we put all the edmtypes which is or can be inherited in the hashset, and if a type is not in hashset it could either mean it is not processed or it could be its not inheritable , so we have to go through that again and again.

Reply:
Right; I'm not suggesting we lazily load which types each type does and does not inherit from -- when asked if type a inherits from type b, we should should load all of the types that type a inherits from and cache that for type a.  Then, if we have an entry for type a, we can just see if type b is in that list.  Type hierarchies are not likely to be deep, so computing the inherited types to the base type should be quick. In fact, for each base type we should check to see if we already have the base types and populate them recursively. 

---
In reply to: [491743487](https://github.com/OData/odata.net/pull/1888#discussion_r491743487) [](ancestors = 491743487)